File: app/src/main/java/com/ftel/demo/MainActivity.java
package com.ftel.demo;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;


import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity {
    private Button pingButton, testButton, chartButton, wifiInfoButton, tracertButton, pageLoadTimerButton, nsLookupButton, speedTestButton, portScanButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 1);
        }

        pingButton = findViewById(R.id.pingButton);
        testButton = findViewById(R.id.testButton);
        chartButton = findViewById(R.id.chartButton);
        wifiInfoButton = findViewById(R.id.wifiInfoButton);
        tracertButton = findViewById(R.id.tracertButton);
        pageLoadTimerButton = findViewById(R.id.pageLoadTimerButton);
        nsLookupButton = findViewById(R.id.nsLookupButton);
        speedTestButton = findViewById(R.id.speedTestButton);
        portScanButton = findViewById(R.id.portScanButton);

        applyGradient(pingButton, "#1DB954", "#191414");
        applyGradient(testButton, "#1DB954", "#191414");
        applyGradient(chartButton, "#1DB954", "#191414");
        applyGradient(wifiInfoButton, "#1DB954", "#191414");
        applyGradient(tracertButton, "#1DB954", "#191414");
        applyGradient(pageLoadTimerButton, "#1DB954", "#191414");
        applyGradient(nsLookupButton, "#1DB954", "#191414");
        applyGradient(speedTestButton, "#1DB954", "#191414");
        applyGradient(portScanButton, "#1DB954", "#191414");

        pingButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), PingActivity.class));
            }
        });

        testButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), ScanWifiActivity.class));
            }
        });

        chartButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), ChartActivity.class));
            }
        });

        wifiInfoButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), WifiInfoActivity.class));
            }
        });

        tracertButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), TracertActivity.class));
            }
        });

        pageLoadTimerButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), PageLoadTimerActivity.class));
            }
        });

        speedTestButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), SpeedTestActivity.class));
            }
        });

        nsLookupButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), NsLookupActivity.class));
            }
        });

        nsLookupButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), NsLookupActivity.class));
            }
        });

        portScanButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), PortScanActivity.class));
            }
        });

    }
    private void applyGradient(Button button, String startColor, String endColor) {
        GradientDrawable gradientDrawable = new GradientDrawable(
                GradientDrawable.Orientation.TOP_BOTTOM,
                new int[]{Color.parseColor(startColor), Color.parseColor(endColor)});
        gradientDrawable.setCornerRadius(8f);
        button.setBackground(gradientDrawable);
    }
}

File: app/src/main/java/com/ftel/demo/MainActivity.kt
package com.ftel.demo

import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ftel.demo.adapter.PingAdapter
import com.ftel.demo.dto.PingResult
import com.ftel.demo.util.PingUtil
import java.io.IOException

import java.net.InetSocketAddress

import java.net.Socket
import kotlin.math.floor


class MainActivity : AppCompatActivity() {
    private lateinit var editTextIPAddress: EditText
    private lateinit var buttonPing: Button
    private lateinit var recyclerView: RecyclerView

    private val avgResponseTimeTextView: TextView by lazy {
        findViewById<TextView>(R.id.avgResponseTimeTextView) ?: throw IllegalStateException("avgResponseTimeTextView not found")
    }
    private val buttonActivity_2: Button by lazy {
        findViewById<Button>(R.id.activity_2Button)
    }
    private lateinit var pingAdapter: PingAdapter

    private val responses: ArrayList<String> = ArrayList()
    private val pingResults : ArrayList<PingResult> = ArrayList()

    private val handler = Handler(Looper.getMainLooper())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        editTextIPAddress = findViewById(R.id.editTextIPAddress)
        buttonPing = findViewById(R.id.buttonPing)
        recyclerView = findViewById(R.id.recyclerViewResult)


        pingAdapter = PingAdapter()

        recyclerView.adapter = pingAdapter
        recyclerView.layoutManager = LinearLayoutManager(this)

        buttonPing.setOnClickListener {
            val ipAddress = editTextIPAddress.text.toString().trim()
            handler.postDelayed(pingTaskRunnable(ipAddress), 0)
        }

        buttonActivity_2.setOnClickListener {
            val intent = Intent(this, Activity_2::class.java)
            startActivity(intent)
        }
    }

    private fun pingTaskRunnable(ipAddress: String): Runnable {
        return Runnable {
            PingTask().execute(ipAddress)
        }
    }

    private inner class PingTask : AsyncTask<String, Void, PingResult>() {
        override fun doInBackground(vararg params: String?): PingResult {
            val ipAddress = params[0] ?: ""
            val pingUtil = PingUtil(ipAddress)
            val ttl = 59
            val timeout = 10000
            val packetSize = 32
            var result : PingResult = pingUtil.sendPing(ttl, timeout, packetSize)
            pingResults.add(result)
            return result
        }

        override fun onPostExecute(pingResult: PingResult) {
            val resultString = """
            Ping #${responses.size + 1}
            Status: ${pingResult.replyStatus}  -  IpAddress: ${pingResult.ipAddress}
            Response Time: ${pingResult.responseTime}ms
        """.trimIndent()

            responses.add(0, resultString)

            val avgResponseTime = pingResults.map { it.responseTime }.average()
            avgResponseTimeTextView.text = "Avg: ${(floor(avgResponseTime * 100)/100).toString()}"
            pingAdapter.clearData()
            pingAdapter.setData(responses)

            // Schedule the next ping after a delay of 1000 milliseconds
            handler.postDelayed(pingTaskRunnable(editTextIPAddress.text.toString().trim()), 1000)
        }


    }

}
